lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -82,9 +87,20 @@ def _check_pos(self, num, name):
     def _proc_upload_reads_params(self, params):
         fwdid = params.get('fwd_id')
         fwdfile = params.get('fwd_file')
-        if not self.xor(fwdid, fwdfile):
-            raise ValueError('Exactly one of a file or shock id containing ' +
+        fwdurl = params.get('fwd_file_url')
+        fwdstaging = params.get('fwd_staging_file_name')
+        if sum(bool(e) for e in [fwdid, fwdfile, fwdurl, fwdstaging]) != 1:
+            raise ValueError('Exactly one of a file, shock id, staging file name or file url containing ' +
                              'a forwards reads file must be specified')
+        if fwdurl:
+            download_type = params.get('download_type')
+            if not (fwdurl and download_type):
+                raise ValueError('Both download_type and fwd_file_url must be provided')
+            fwdfile = self._download_web_file(fwdurl, download_type)
+        
+        if fwdstaging:
+            fwdfile = self._download_staging_file(params.get('user_id'), fwdstaging)
+
```

Comment:
A possible issue with downloading the files here and having the rest of the code treat them as if they already existed locally is that the error messages you get back from problems later in the code might not be helpful - it'll just include the file name of the downloaded file, and not provide information regarding how to find the source of the file. If you look at the error handling in the main method below the errors are different based on whether the file source was in shock or from a local file; my feeling is that something similar needs to be done for the alternate file sources.